docs/design/features/data-stream.md

```diff
@@ -0,0 +1,106 @@
+
+### Reader usage
+
+The reader should first define a series of type names that it is able to interpret and consume. These type names should match the names defined by the target. These names could be used to map types in the target process with their numeric ID and version. The reader is not expected to have any hardcoded numeric type IDs as these are subject to change between target versions.
+
+The reader is expected to be resiliant in recieving an unknown version of a type and gracefully interpret it. Two examples of graceful interpretation are describing it as `"Unknown"` or printing its memory address in the target process.
+
+After acquiring the target process's `data_stream_context_t` address, it should be validated and endianness computed.
+
+The first time the data stream is read, the target processes types are enumerated in reserve chronological order (Last In/First Out) and this data may be used throughout the lifetime of the target process. During the enumeration of types the following can be done:
+
+* Creation of a fast mapping from name to numeric ID.
+
+* Creation of look-up map to type details (e.g., field offsets).
+
+* Validation of supported type versions.
+
+After type enumeration is complete, instance streams can be enumerated and interpreted. The size contained within the type description allows the reader to read in the entire type and then use field offsets to poke into that memory. The reading in of the entire data type helps with efficiency and acts as a versioning resiliance mechanism. Adding new fields to a type, without changing the version, need not represent a breaking change.
+
+### Design FAQs
+
+---
+
+**Q1** Why are streams allowed to grow?
+
+**A1** Consider the case where a data structure in the target process has a specific use case but the reader has either stricter or looser requirements. An example would be a thread pool used in the target process. This structure would ideally only be concerned with current threads in the target process, exited threads having been removed. However, the reader process likely has a need for knowing when a thread instance has exited to update its own internal state. A possible solution is to fully query the thread pool data structure each time. However, if instead entries for created and deleted threads are recorded in a stream, the reader only needs to know the delta as opposed to querying the thread pool each time. The logic follows for any data structure that contains objects with transient lifetimes.
+
+---
+
+**Q2** Why are the contents of a stream immutable?
+
+**A2** Having streams that are mutable means the reader _must_ always re-read the full stream to validate for updates. If the contents of a stream are instead immutable _and_ in reverse chronological order (LIFO), then entries for "deleted" or "invalidated" data are possible, which enables readers to consume deltas and reduce cross-process inspection.
+
```

Comment:
At least some of these data streams need to included in the minidump payload. How are we going to decide which data streams need to be included in the minidump payload?

Also, we need to keep an eye on the overhead that the data streams add to the raw minidump size so that it does not get out of control.